File: src/easy/reverse-linked-list/1.ts
import { ListNode } from './list-node';
/*
 * @lc app=leetcode id=206 lang=typescript
 *
 * [206] Reverse Linked List
 */

// @lc code=start

// Definition for singly-linked list.
function reverseList(head: ListNode | null): ListNode | null {
  const nums: number[] = [];
  while (head) {
    nums.push(head.val);
    head = head.next;
  }

  let result: ListNode | null = null;
  for (let i = 0; i < nums.length; i++) {
    const tempNode = new ListNode(nums[i], result);
    result = tempNode;
  }
  return result;
}

// space-complexity: O(n) time-complexity: O(n) pretty sad :(
// @lc code=end

File: src/easy/merge-two-sorted-lists/1.go
/*
 * @lc app=leetcode id=21 lang=golang
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
package mergeTwoLists

import "sort"

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	ls := []int{}

	for list1 != nil {
		ls = append(ls, list1.Val)
		list1 = list1.Next
	}

	for list2 != nil {
		ls = append(ls, list2.Val)
		list2 = list2.Next
	}

	sort.Ints(ls)

	var last *ListNode
	for i := len(ls) - 1; i >= 0; i-- {
		current := &ListNode{
			Val:  ls[i],
			Next: last,
		}
		last = current
	}

	return last
}

// @lc code=end

// time complexity: O(n + m) space complexity: O(n + m)
